src/hotspot/share/gc/g1/g1HeapRegion.cpp

```diff
@@ -624,6 +623,7 @@ class G1VerifyLiveAndRemSetClosure : public BasicOopIterateClosure {
     _num_failures(0) { }
 
   void set_containing_obj(oop const obj) {
+    assert(!G1CollectedHeap::heap()->is_obj_dead_cond(obj, _vo), "Precondition");
     _containing_obj = obj;
```

Comment:
Sounds like a reasonable thing to do. I'll move that assert as well.

Reply:
You mean moving `assert(_containing_obj != nullptr, "must be");` from `do_oop_work`? I think it should stay there, as it verifies that someone called `set_containing_obj` before doing `do_oop_work`?

Reply:
I think it is better to add a new one insteading of moving it.

Reply:
Moving it removed yet another 0.5s from the runtime of the test. Are you sure you want to keep it?

FWIW, this is the only (current) usage of set_containing_obj and oop_iterate over the closure. I don't think it is a great risk that someone will accidentally move/copy the oop_iterate call and fail to call set_containing_obj.
```
      cl.set_containing_obj(obj);
      obj->oop_iterate(&cl);
```

Reply:
I think it should be moved; the risk of misuse here is low.

Reply:
> I think it should be moved; the risk of misuse here is low.

If we move it, could/should we add some related comments in the class `G1VerifyLiveAndRemSetClosure` or the related methods?

Reply:
I'd prefer to have a clear assert failure instead of SEGV. Maybe if we move the assert closer to its first use (`LiveChecker<T> live_check(this, _containing_obj, p, obj, _vo);`), the null-check would filter enough to gain the same perf improvement?

Reply:
I have an alternative that removes the set_containing_obj and instead sets it in the constructor. This way it should be enough to assert in the constructor. Please, take a look at this patch:
https://github.com/openjdk/jdk/compare/master...stefank:jdk:8329570_is_obj_dead_cond_2

Reply:
With the new G1VerifyFailureCounter in the _2 patch, we could even pull this a bit further and remove the other_failures counter:
https://github.com/openjdk/jdk/commit/8ee0b46f64bb609b4c0473fb35df47908bbeced9
https://github.com/openjdk/jdk/compare/master...stefank:jdk:8329570_is_obj_dead_cond_3
:shrug:

Reply:
I was think of something similar, also makes the iteration actually stop at `G1MaxVerifyFailures` instead of at some related but arbitrary larger number.

Reply:
Honestly, I think we are overthinking this. The original patch gives us a major bang for the buck, and it has a benefit of being obvious and thus very backportable. I suggest we go with the current version, and then do any other followups.

Reply:
I can clarify my position. I completely agree with @shipilev, hence my approval of still stands. I think the current fix is a very reasonable solution to the problem.

But my initial thought / reaction when looking at the code was "why is the closure have mutable state" and "the check agains `G1MaxVerifyFailures` looks strange".

Reply:
Yeah, I think the suggestion to keeping the null-check in do_oop_work was overthinking things. But, since that seemed important I proposed an alternative solution that I think actually is cleaner and clearer than what the current code was doing